### PR TITLE
First pass at using compressed format when uploading textures

### PR DESCRIPTION
**Pull request by samcake**
- THis work DO NOT ENALBE the compression of textures but prepare for it.
- Provide compressed format for all the textures uploaded to GPU in order to save memory. 
- Do not use compression on normal map yet
- Added a counter of the amount of texture bound as resource during a frame
- added a "TextureGPUVirtualMemoryUsage" counter in Texture in order to monitor the size expected to be taken by a texture if not compressed
 This is to be compared with the actual true value reported by TextureGPUMemoryUsage which is taking it s value from the size reported by opengl.
- updated the renderStats,js qml to reflect these 2 new counters.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9d1f91fa19da870952f556778a1fffde8a0ee571 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9404846ca4e322c9d4ffc564356b89c0e697b96a available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 582a6b7db308f9bd2ea9b0022254d475deffa101 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by kevinmthomas-carpool** [comment]
Testing using a NVIDIA GeForce GTX 970
Testing on Music Domain

Values from TechPowerUp GPU-Z
Memory Size 4096 MB
GPU Memory Used Average 1942 MB 
GPU Load 9%

Values from renderstats.js
Num Buffers 2021
CPU 1819
GPU 793
gpu::Buffer Memory 34.5 Mb
CPU 31.0 Mb
GPU 27.9 Mb
Num Textures 973
CPU 561
GPU 555
Frame 876
gpu::Texture Memory 2204.6 Mb
CPU 17.1 Mb
GPU 1152.8 Mb
GPU Virtual 1984.2 Mb
Triangles 891 K
Triangles 802 K
rate 13 MT/s
Drawcalls 520
API Drawcalls 452
GPU Drawcalls 460
rate 7 K/s
Items 198
Opaques 178
Translucents 17
Lights 0

I can confirm that the value of GPU is 1152.8 Mb and the GPU Virtual 1984.2 Mb.

I can report I see no loss in quality of any kind any of the textures throughout my domain compared to the master.

**Comment by samcake** [comment]
The compression is doing what we expect correctly but we see a big wait on the first gpu upload..
We are going to try to generate the compressed version inthe same thread doing the NetworkTexture load to avoid to have to pay that price in the main thread.

**Comment by thoys** [comment]
:+1: will QA after that change

**Comment by TNS-TrueNorth** [comment]
This PR populates 2 instances of Sandbox in TaskManager (see screenshot) and neither of them appear in the Notifications bar in the bottom right corner.

**Comment by TNS-TrueNorth** [comment]
Testing using a NVIDIA GeForce GTX 980 on Win10
Testing on Playa Domain

Values from TechPowerUp GPU-Z
 Memory Size 4096 MB
 GPU Memory Used Average 875 MB 
 GPU Load 20%

Values from renderstats.js (see attached log)
[hifi-log_192.168.1.33_2016-04-01_10.44.02.txt](https://github.com/highfidelity/hifi/files/200135/hifi-log_192.168.1.33_2016-04-01_10.44.02.txt)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1ca87fa06962ae3d8645a8665f316492f6926d1f available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: be38de32398aac789757b21347ec1479590d744b available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f3e6c490a4386f02b37db1eee09201fa429d1925 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by ZappoMan** [comment]
To TEST:
Look at the value from GPU VIrtual Memory vs GPU Memory and tell me that the GPUMemory returned is much lower and also match approximately the amount of memory reported by tool like GPU-Z
What about the quality, do you see a difference compared to Master ?

**Comment by OmegaHeron** [comment]
Windows 7 - GTX660M - 2GB GPU RAM

Before starting interface GPU-Z reports 165M of 2G RAM used
Very simple scene at Heron - 4 textures + AV 360M used.  Since I've stared at its ground texture for > year I can say it looks same as always.
Playa - 1422M fully loaded - 2 Avatars present beyond my own.
Returning to Heron - GPUZ 465M

Production build (4618) - GPU-Z reports 165M used
Very simple scene at Heron - 4 textures + AV 399M used.
Playa - 1969MM used - Same 2 Avatars + my own seen same as PR test
Returning to Heron -554M used

I can't really see a difference in textures at Playa between production and this PR, but, Playa is hell on my systems and it's hard to move around and really inspect things up close where I might see a difference.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 710c4a1b8f5fa01e59cff944e2531b5b4cf11faf available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9bc06d92d934968e39585aa4759bd629de1297c available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

File: libraries/model/src/model/TextureMap.cpp [hunk 1]
```diff
@@ -131,22 +182,15 @@ gpu::Texture* TextureUsage::createNormalTextureFromNormalImage(const QImage& src
             image = image.convertToFormat(QImage::Format_RGB888);
         }
     } else {
-        if (image.format() != QImage::Format_ARGB32) {
-            image = image.convertToFormat(QImage::Format_ARGB32);
-        }
+        image = image.convertToFormat(QImage::Format_RGB888);
```
Comment (jherico) [inline libraries/model/src/model/TextureMap.cpp]: why are you dropping the alpha channel here?

**Comment by hifi-gustavo** [comment]
Build console output for commit: e28e85a29c893db46f3314deb2c8d8e39fee8fb5 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 16174df0f319f7cda70ee2ec28fb87057f006a60 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 92df0e45e69e5e906f5411526e20deead287024a available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a6bac7ad927789932f532a546ae27680fe626f99 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cf983e627dd341b683ed3cb8a58c834233572c4a available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 76dd7405787754c1a9c11ab702a52151fd95eb81 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 04e71df3bcc3d09b338fd81681d2a4e00c4a1f87 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: be178e9e0988a81076bc091bf38a45ee9a53ccf7 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4374bd26ae210b448b3763f87038e6e950c252d4 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.dmg

**Comment by samcake** [comment]
THis PR is now ready for Review and merge.
This PR DO NOT enable compression of texutres because this introduce a stuttering on the rendering thread. Instead this PR prepare for it, the goal beeing to find a way to generate the compression bytes without relying on the openGL compression upload functions as is.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ef70d41cf47a0f14b5b9539c380ae353ba832064 available at: https://gist.github.com/6a4abd0d89ce5ffe7219ee7dd41ea130<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7536.dmg

File: libraries/gpu/src/gpu/GLBackendTexelFormat.cpp [hunk 2]
```diff
@@ -0,0 +1,426 @@
+//
+//  Created by Bradley Austin Davis on 2016/04/03
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexelFormat.cpp]: this header seems wrong. Did Austin write this? or Sam?

File: libraries/gpu/src/gpu/GLBackendTexelFormat.cpp [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    GLTexelFormat texel = { GL_RGBA, GL_RGBA, GL_UNSIGNED_BYTE };
+    return texel;
+}
+
+GLTexelFormat GLTexelFormat::evalGLTexelFormat(const Element& dstFormat, const Element& srcFormat) {
+    if (dstFormat != srcFormat) {
+        GLTexelFormat texel = { GL_RGBA, GL_RGBA, GL_UNSIGNED_BYTE };
+
+        switch (dstFormat.getDimension()) {
+        case gpu::SCALAR: {
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexelFormat.cpp]: coding standard: formatting is wrong... https://readme.highfidelity.com/docs/coding-standard#section-4-1-9-switch-case-statements-

File: libraries/gpu/src/gpu/GLBackendTexelFormat.cpp [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+        default:
+            qCDebug(gpulogging) << "Unknown combination of texel format";
+        }
+        return texel;
+    } else {
+        GLTexelFormat texel = { GL_RGBA8, GL_RGBA, GL_UNSIGNED_BYTE };
+
+        switch (dstFormat.getDimension()) {
+        case gpu::SCALAR: {
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexelFormat.cpp]: https://readme.highfidelity.com/docs/coding-standard#section-4-1-9-switch-case-statements-

File: libraries/gpu/src/gpu/GLBackendTexelFormat.cpp [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+                texel.format = GL_RGB;
+                // the type should be float
+                texel.internalFormat = GL_R11F_G11F_B10F;
+                break;
+
+            case gpu::DEPTH:
+                texel.format = GL_DEPTH_COMPONENT; // It's depth component to load it
+                texel.internalFormat = GL_DEPTH_COMPONENT32;
+                switch (dstFormat.getType()) {
+                case gpu::UINT32:
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexelFormat.cpp]: https://readme.highfidelity.com/docs/coding-standard#section-4-1-9-switch-case-statements-

File: libraries/gpu/src/gpu/GLBackendTexture.cpp [hunk 6]
```diff
@@ -49,49 +48,145 @@ const GLenum GLBackend::GLTexture::CUBE_FACE_LAYOUT[6] = {
 
 // Create the texture and allocate storage
 GLBackend::GLTexture::GLTexture(const Texture& texture) : 
-    _storageStamp(texture.getStamp()), _texture(allocateSingleTexture()),
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexture.cpp]: did you mean to remove `_texture`

**Comment by samcake** [comment]
To TEST:
Do a smoke test in different domains, the rendering should still be very smooth while loading new content.
Look at the value from GPU VIrtual Memory and GPU Memory displayed with RenderStats.js (the one from this PR), they should be the same.
AS you come to new area of a domain, the number of texture in "transfer" should go up and always come back to 0 once everything has been loaded
the GPUMemory value should match approximately the amount of memory reported by tool like GPU-Z

File: libraries/gpu/src/gpu/GLBackendTexelFormat.cpp (hunk 2 above)
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackendTexelFormat.cpp]: THis file is just a move from the GLTexelFormat.h that was itself a pull off from the code from GLBackendTexture.cpp.
So i kept the separation, gave credit to Austin for doing the split, and move it to a cpp file witha naming that matches better the gpu lib file names.

File: libraries/gpu/src/gpu/GLBackendTexture.cpp (hunk 6 above)
Comment (samcake) [inline libraries/gpu/src/gpu/GLBackendTexture.cpp]: no the new code path keep _texture to 0 until it is fully defined further after the full gl creation and if needed the full gl transfer.
the _texture field is finaly assigned in the postTransfer() call

**Comment by OmegaHeron** [comment]
@samcake - When you say "the GPUMemory value should match approximately the amount of memory reported by tool like GPU-Z" - Clarification needed.  Prior to launching Interface I see GPU-Z indicating 158M used - then I run Interface and see it showing 364.  Interface reports gpu::Texture Memory @ 137.8 - Now, if I add the GPU RAM in use prior to Interface + Interface's reported use I'm sort of close to matching, but, if gpu::Texture Memort in renderStats.js should be (nearly) same as GPU-Z then it's most certainly not.  So - let me (us) know how to interpret that please.

**Comment by samcake** [comment]
@OmegaHeron Yes, that s what i would call "close enough"

The  true Memory allocated by the driver measured by GPU-Z is what it is. we cannot know exactly how much is that because it's driver and platform dependent.
That being said, the reported values for GPU Texture Memory And GPU Buffer Memory is what we can count and measure on the app side.
In your case gpu-z reports ~ 210Mb and you measure ~150Mb (adding 138Mb you reported and guessing about 12Mb of geometry, correct me if i m wrong).
So this Seems about right.

One more thing that we need to measure more precisely is the consumption from the QML and Overlay elements which are not taken into account. 

THe other aspect which is important to check is the dynamic behavior of INteface and the memory consuption as you navigate the world from domains to domains.
Each time you transition from one domain to the next, we should clear out most of the GPU memory (except for the rendering textures ).
GPU-Z should reflect this behavior too.

Let me know if you have any other questions.
I will add up on my to do a counter for the GPU memory that should be accounted to QML & 2D overlays

**Comment by OmegaHeron** [comment]
@samcake - Thanks, now I can move along with testing.

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**PR Build:** 7536
**Testing Architecture:** 64-bit PC Desktop
**Testing OS:** Windows 10
**Testing Locations:** Sandbox, Playa, Cellscience, Music
**Smoke Test Results:** Partially Successful test.

                                  |Playa    |Cellscience |Music  
----------------------------------|---------|------------|--------
Render Rate:                      |64       |61          |63           
Present Rate:                     |60       |60          |60          
Simrate:                          |64       |62          |63           
Render Stats Num Buffers:         |4272     |2576        |2009              
Render Stats gpu::Buffer Memory:  |98.6Mb   |32.6Mb      |31.9Mb              
Render Stats Num Textures:        |513      |212         |1282              
Render Stats gpu::Texture Memory: |1550.0Mb |486.4Mb     |3572.0Mb              
Render Stats Triangles:           |582k     |571k        |965k            
Render Stats Drawcalls:           |25593    |5493        |623             
Render Stats Items:               |272      |296         |287  
GPU-Z Memory Size                 |4096Mb   |4096Mb      |4096Mb
GPU-Z Initial Memory		  |250Mb    |250Mb	 |250Mb
GPU-Z Memory Used		  |2395Mb   |1324Mb      |3296Mb
GPU-Z Memory Load		  |9%       |9%          |12%

**sandbox ‘createtestblocks.js’ Result:** Failure as the entities physics were unstable random in their behavior.
**sandbox Import FBX Result:** Successful test.
**sandbox ‘harmonicOscillator.js’ Result:** Successful test.
**sandbox Echo Server Audio Result:** Successful test originally.  Upon visiting the Playa domain and in contact with @Thoys it became robotic.  Relogged and worked normal so the robotic audio is most likely not tied to the PR.  Successful test on other domains represented in this analysis.
**sandbox ‘reverbTest.js’ Result:** Successful test.
**Additional Test Results Requested:** Successful test.  There appeared to be a smooth transition between domains.  I can confirm the GPU texture memory gets cleared out from domain to domain as it has diminished from visiting Cellscience from Playa.  GPU memory increased dramatically when visiting Music but that is expected behavior as there are a very large amount of large-sized textures.
**Successful Logout/Crash:** Successful logout.

**Comment by OmegaHeron** [comment]
My framerate (even in sandbox or playa) remain stable while loading :+1: 
In all tests made I never saw a difference in GPU vs GPU Virtual :+1: 
Textures in transfer always settle back to 0 :+1: 
Other tests match Kevin's report.

**Pull request merged**